### PR TITLE
experimental enable jemalloc API for FreeBSD too.

### PR DESCRIPTION
**Pull request by devnexen**
Contrary to other systems, jemalloc is part of the FreeBSD's libc.
As such APIs are directly exposed. Enabling super pages
as well, supporting same page size.

### CODE REVIEW

File: folly/experimental/JemallocHugePageAllocator.cpp [hunk 1]
```diff
@@ -40,7 +44,7 @@ bool folly::JemallocHugePageAllocator::hugePagesSupported{true};
 #define MALLOCX_TCACHE_NONE 0
 #define MADV_HUGEPAGE 0
 
-#if !defined(JEMALLOC_VERSION_MAJOR) || (JEMALLOC_VERSION_MAJOR < 5)
+#if (!defined(JEMALLOC_VERSION_MAJOR) || (JEMALLOC_VERSION_MAJOR < 5))
```
Comment (yfeldblum) [inline folly/experimental/JemallocHugePageAllocator.cpp]: This bit is not related.

File: CMake/FollyConfigChecks.cmake [hunk 2]
```diff
@@ -20,7 +20,11 @@ include(CheckSymbolExists)
 include(CheckTypeSize)
 include(CheckCXXCompilerFlag)
 
-CHECK_INCLUDE_FILE_CXX(jemalloc/jemalloc.h FOLLY_USE_JEMALLOC)
+if (NOT CMAKE_SYSTEM_NAME STREQUAL "FreeBSD")
+  CHECK_INCLUDE_FILE_CXX(jemalloc/jemalloc.h FOLLY_USE_JEMALLOC)
+else()
+  CHECK_INCLUDE_FILE_CXX(malloc_np.h FOLLY_USE_JEMALLOC)
+endif()
```
Comment (yfeldblum) [inline CMake/FollyConfigChecks.cmake]: Should switch the conditions around. That way, further conditions can easily be added as necessary.

```
if(CMAKE_SYSTEM_NAME STREQUAL "FreeBSD")
  ...
else()
  ...
endif()
```

File: folly/experimental/JemallocHugePageAllocator.cpp [hunk 3]
```diff
@@ -21,14 +21,18 @@
...
 #include <sstream>
 
-#if defined(MADV_HUGEPAGE) && defined(FOLLY_USE_JEMALLOC) && !FOLLY_SANITIZE
+#if (defined(MADV_HUGEPAGE) || defined(MAP_ALIGNED_SUPER)) && defined(FOLLY_USE_JEMALLOC) && !FOLLY_SANITIZE
+#if !defined(__FreeBSD__)
 #include <jemalloc/jemalloc.h>
-#if (JEMALLOC_VERSION_MAJOR >= 5)
+#else
+#include <malloc_np.h>
+#endif
```
Comment (yfeldblum) [inline folly/experimental/JemallocHugePageAllocator.cpp]: Likewise.

File: folly/experimental/JemallocHugePageAllocator.cpp [hunk 4]
```diff
@@ -21,14 +21,18 @@
...
+#if defined(__FreeBSD__) || (JEMALLOC_VERSION_MAJOR >= 5)
```
Comment (yfeldblum) [inline folly/experimental/JemallocHugePageAllocator.cpp]: Why do we need the `__FreeBSD__` check here? Does jemalloc on FreeBSD not define `JEMALLOC_VERSION_MAJOR`?

File: folly/portability/Malloc.h [hunk 5]
```diff
@@ -28,8 +28,12 @@
 #endif
 // JEMalloc provides it's own implementation of
 // malloc_usable_size, and that's what we should be using.
+#ifndef __FreeBSD__
```
Comment (yfeldblum) [inline folly/portability/Malloc.h]: I prefer `#if !defined(...)`, since that way it is easier to add clauses to the condition as necessary.

File: folly/experimental/JemallocHugePageAllocator.cpp (hunk 4 above)
Comment (devnexen) [reply folly/experimental/JemallocHugePageAllocator.cpp]: Nope. they keep all "close to the vest". fair enough for the rest.

File: folly/experimental/JemallocNodumpAllocator.h [hunk 6]
```diff
@@ -24,7 +24,11 @@
 #if defined(FOLLY_USE_JEMALLOC) && !FOLLY_SANITIZE
 
 #include <folly/portability/SysMman.h>
+#if !defined(__FreeBSD__)
```
Comment (yfeldblum) [inline folly/experimental/JemallocNodumpAllocator.h]: Same.

File: folly/memory/Malloc.h [hunk 7]
```diff
@@ -29,8 +29,12 @@
  */
 #if (defined(USE_JEMALLOC) || defined(FOLLY_USE_JEMALLOC)) && !FOLLY_SANITIZE
 // We have JEMalloc, so use it.
+#if !defined(__FreeBSD__)
```
Comment (yfeldblum) [inline folly/memory/Malloc.h]: Same.

File: folly/portability/Malloc.h [hunk 8]
```diff
@@ -28,8 +28,12 @@
 #endif
 // JEMalloc provides it's own implementation of
 // malloc_usable_size, and that's what we should be using.
+#if !defined(__FreeBSD__)
```
Comment (yfeldblum) [inline folly/portability/Malloc.h]: Same.

**Comment by yfeldblum** [comment]
There was some duplication here, the root cause of which I address in d054e34e13dd9c41005d234b7b1d4b115927f92d (on master). Please rebase past that and resolve conflicts?
